### ISSUE TITLE
Indicators which have two inputs with the same name are unusable

### ISSUE THREAD

**Issue by cirla** [cirla/tulipy#1]
Many indicators (e.g. `ti.add`) have two inputs both named `'real'`, which renders them unusable by the current bindings (unless, of course, you want to use the same data for both inputs).

**Comment by codeplea** [comment]
The Tulip Indicators C interface determines inputs by order, so the names don't matter there at all. When I added the `input_names`, it was for a hint to charting programs to be able to automatically set `open`, `high`, `low`, `close`, and `volume` for the indicators that only make sense with those. `real` became a catch-all for inputs that could make sense taking any type of input, basically an indication that the charting program should allow the user to select anything for that input. In hindsight, perhaps I should have called it `input_types` instead of `input_names`.

Would it be un-pythonic to pass in arguments by order in an array instead of an object? E.g.

```python
ti.xxx({'open': DATA1, 'close': DATA2}, {'period': 5.0})
```

would become:

```python
ti.xxx([DATA1, DATA2], [5.0])
```

That is how I did the bindings for node.js over [here](https://github.com/TulipCharts/tulipnode).

I don't use python enough to know if this would be unidiomatic usage. What do you think?

**Comment by cirla** [comment]
Ideally for idiomatic use, each signature would match the function, e.g.:
```python
def xxx(open, close, period):
```
and they could be called with either positional args or keyword args, e.g.:
```python
ti.xxx(DATA1, DATA2, 5.0)
# or
ti.xxx(open=DATA1, close=DATA2, period=5.0)
```
but I was trying to avoid generating (or manually writing) the signatures for each indicator so that I could just write a single `__call__` implementation thanks to the consistent interface between all indicators in the library implementation. This way would still run into the same problem as the current implementation, however, unless I renamed the arguments for these cases to something `real1` and `real2`.

In short, the current implementation was done out of laziness and trying to take advantage of all of the C indicator functions having the same interface while still trying to have an air of idiomatic python by allowing for named keys in a dict. It may be best to be consistent with the official node.js bindings and go for a python interface like:
```python
ti.xxx([DATA1, DATA2], [5.0])
# or
ti.xxx(inputs=[DATA1, DATA2], options=[5.0])
```
and returning a list of outputs instead of a dict as the `.inputs`, `.options`, and `.outputs` properties make clear the required ordering.

**Comment by codeplea** [comment]
I've thought about this a bit, and I'm tempted to change Tulip Indicators and get rid of "real". I think it makes more sense to call it "input1", "input2", etc. What are your thoughts on that?

I agree it would be nice for Tulipy to just take the arguments in a straight-forward way. My thought behind Tulip Indicators was that people wouldn't call one indicator, they would call all the indicators. So they Tulip Indicators interface reflects that.

I think with Tulipy, however, end-users will use it in scripts where they actually do want to call only one indicator. For example, it would be nice if a crossover strategy could be coded by an end user as:

```Python
buy = ti.crossover(ti.sma(close,5), ti.sma(close,24))
```

I think it should be pretty easy to write a script to generate a python wrapper to make that happen. It should still follow the same call path, though, to get the error handling and input trimming. I'm working on this now. I'll post it in a minute.

**Comment by codeplea** [comment]
Ok, it's a mess, and I probably did some things wrong (I'm not a Python programmer), but I added a branch that lets you call with something like the above. e.g.

```Python
ti.sma(np.array([5.0,1.0,2.0,...]), 5.0)
```

See [here](https://github.com/cirla/tulipy/tree/calling_convention).

Building is a two-step process. First you build with `make src/tulipy.pxd` to generate the interface. This runs `interface.py` which generates a function for each indicator.

There is probably a better way to do it. Also, I have an indicator function called `min` which clashes with the Python `min` function. I'm sure there are other namespace clashes too. I'm not sure how to handle it.

**Comment by cirla** [comment]
I tried something based off your work [in this branch](https://github.com/cirla/tulipy/compare/gen?expand=1) (see [here](https://github.com/cirla/tulipy/blob/gen/README.ipynb) for example use) and it seems to work very well, though it generates a second Python package named `tulipy-gen` which I'd like to avoid, increases the compile time quite a bit, and adds a lot of `_Indicator_<name>` classes to the module namespace.

**Comment by cirla** [comment]
I may just move the gen phase out of setup.py and make it a separate make target, then check the generated file into source control as well since it's tied to the TI header and amalgamation anyway.

**Comment by codeplea** [comment]
I think that's best. That what I did in [my branch](https://github.com/cirla/tulipy/tree/calling_convention). It's one less thing to go wrong for the end-user.

**Comment by codeplea** [comment]
Have you considered packaging a pure Python wrapper of the Cython extension? I think then we could do something like this and avoid all the classes:

```Python
def sma(real, period):
    return tulipy_raw.sma([real], [period])
sma.name = "sma"
sma.full_name = "Simple Moving Average"
sma.inputs = ...
```

**Comment by codeplea** [comment]
Added a new branch: [front](https://github.com/cirla/tulipy/tree/front). It puts the Cython wrapper into `tulipy_raw`, and then has a pure python wrapper for `tulipy`. I think one advantage, is that it leaves a clean interface to the end-user.

**Comment by cirla** [comment]
See #6 which seems to work well

**Comment by codeplea** [comment]
Looks good to me. Nice work!
